File: solution/python3/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        brackets = {
            ')': '(',
            '}': '{',
            ']': '[',
        }

        for c in s:
            if c in ['(', '{', '[']:
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False

                if stack.pop() != brackets[c]:
                    return False

        if len(stack) != 0:
            return False

        return True


# @lc code=end

File: solution/java8/20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Stack<String> stack = new Stack<String>();
        String[] chars = s.split("");
        Map<String, String> parentheses = new HashMap<>();
        parentheses.put("(", ")");
        parentheses.put("{", "}");
        parentheses.put("[", "]");

        for (String c: chars) {
            if (parentheses.containsKey(c)) {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                String opener = stack.pop();
                if (c.equals(parentheses.get(opener))) {
                    continue;
                } else {
                    return false;
                }
            }
        }

        return stack.isEmpty();

    }
}
// @lc code=end

